src/auspex/plotting.py

```diff
@@ -6,66 +6,78 @@
-            logger.info("Using existing Bokeh server")
-            return
-        logger.info("Starting Bokeh server")
-        args = ["bokeh", "serve", "--port", "5006", "--allow-websocket-origin=localhost:8888", "--allow-websocket-origin=localhost:8889", "--allow-websocket-origin=localhost:8890"]
-        self.p = subprocess.Popen(args, env=os.environ.copy())
-        self.write_session_pid()
-        # sleep to give the Bokeh server a chance to start
-        # TODO replace this with some bokeh client API call that
-        # verifies that the server is running
-        time.sleep(3)
+    def __init__(self, plot_desc={}, status_port = 7771, data_port = 7772):
+        super(MatplotServerThread, self).__init__()
+        self.plot_desc = plot_desc
+        self.status_port = status_port
+        self.data_port = data_port
+        self.daemon = True
+        self.stopped = False
+        self.start()
+
+    async def poll_sockets(self):
+        while not self.stopped:
+            evts = dict(await self.poller.poll(50))
+            if self.status_sock in evts and evts[self.status_sock] == zmq.POLLIN:
+                ident, msg = await self.status_sock.recv_multipart()
+                print("Got {} from {}".format(msg.decode(), ident.decode()))
+                if msg == b"WHATSUP":
+                    await self.status_sock.send_multipart([ident, b"HI!", json.dumps(self.plot_desc).encode('utf8')])
+            await asyncio.sleep(0.010)
+
+    async def _send(self, name, data, msg):
+        md = dict(
+            dtype = str(data.dtype),
+            shape = data.shape,
+        )
+        await self.data_sock.send_multipart([msg.encode(), name.encode(), json.dumps(md).encode(), data])
 
-    def terminate(self):
-        if self.p:
-            print("Killing bokeh server process {}".format(self.p.pid))
+    def send(self, name, data, msg="data"):
+        self._loop.create_task(self._send(name, data, msg=msg))
+
+    def stop(self):
+        self.stopped = True
+        pending = asyncio.Task.all_tasks(loop=self._loop)
+        self._loop.stop()
+        time.sleep(1)
+        for task in pending:
+            task.cancel()
             try:
-                for child_proc in psutil.Process(self.p.pid).children():
-                    print("Killing child process {}".format(child_proc.pid))
-                    child_proc.terminate()
-            except:
-                print("Couldn't kill child processes.")
-            self.p.terminate()
-            self.p = None
-            os.remove(self.pid_filename)
+                self._loop.run_until_complete(task)
+            except asyncio.CancelledError:
+                pass
+        self._loop.close()
 
-    def write_session_pid(self):
-        with open(self.pid_filename, "w") as f:
-            f.write("{}\n".format(self.p.pid))
+    def run(self):
+            self._loop = zmq.asyncio.ZMQEventLoop()
+            asyncio.set_event_loop(self._loop)
```

Comment:
The nice thing about the asyncio event loop is that the `pyZMQ` event loop is just a thin wrapper around it that shouldn't add much overhead. I think we'd have to write our own ZMQ-socket-friendly `uvloop` to use it.